src/actions/requests.rs

```diff
@@ -217,69 +207,76 @@ impl<'a> Action<'a> for Definition {
         let file_path = parse_file_path!(&params.text_document.uri, "goto_def")?;
         let span = ctx.convert_pos_to_span(file_path.clone(), params.position);
         let analysis = ctx.analysis.clone();
         let vfs = ctx.vfs.clone();
 
-        let compiler_handle = thread::spawn(move || {
-            let result = analysis.goto_def(&span);
-
-            t.unpark();
-
-            result
+        let (sender, receiver) = mpsc::channel();
+        let mut expecting_to_receive = 1;
+        let compiler_sender = sender.clone();
+        WORK_POOL.spawn(move|| {
+            let _ = compiler_sender.send(ResponseFrom::Compiler(analysis.goto_def(&span)));
         });
 
         // Racer thread.
-        let racer_handle = if ctx.config.lock().unwrap().goto_def_racer_fallback {
-            Some(thread::spawn(move || {
+        if ctx.config.lock().unwrap().goto_def_racer_fallback {
+            WORK_POOL.spawn(move|| {
 
                 let cache = racer::FileCache::new(vfs);
                 let session = racer::Session::new(&cache);
                 let location = pos_to_racer_location(params.position);
 
-                racer::find_definition(file_path, location, &session)
-                    .and_then(location_from_racer_match)
-            }))
-        } else {
-            None
-        };
+                let out = racer::find_definition(file_path, location, &session)
+                    .and_then(location_from_racer_match);
 
-        thread::park_timeout(Duration::from_millis(::COMPILER_TIMEOUT));
+                let _ = sender.send(ResponseFrom::Racer(out));
+            });
+            expecting_to_receive += 1;
+        }
 
-        let compiler_result = compiler_handle.join();
-        match compiler_result {
-            Ok(Ok(r)) => {
-                let result = vec![ls_util::rls_to_location(&r)];
-                trace!("goto_def (compiler): {:?}", result);
-                Ok(result)
+        let start = Instant::now();
+        let max_wait = Duration::from_millis(::COMPILER_TIMEOUT);
+
+        for _ in 0..expecting_to_receive {
+            let elapsed = start.elapsed();
+            if elapsed >= max_wait {
+                break;
             }
-            _ => {
-                match racer_handle {
-                    Some(racer_handle) => match racer_handle.join() {
-                        Ok(Some(r)) => {
-                            trace!("goto_def (Racer): {:?}", r);
-                            Ok(vec![r])
-                        }
-                        Ok(None) => {
-                            trace!("goto_def (Racer): None");
-                            Ok(vec![])
-                        }
-                        _ => {
-                            debug!("Error in Racer");
-                            Ok(vec![])
-                        }
-                    },
-                    None => Ok(vec![]),
+
+            match receiver.recv_timeout(max_wait - elapsed) {
```

Comment:
If I'm not mistaken, this changes the semantics to treat both analysis data and racer as equal and use the first one returned, whereas the previous one preferred the analysis data if possible. Probably not a bad thing, but something to keep in mind

Reply:
That's right. It seemed the only sensible thing to do. Waiting then entire timeout is not something we ever want to happen, its very much a measure to cap the worst case scenario.

I think it would be cleaner just to use one or the other if you want full clarity. Waiting the max request timeouts for the racer results hardly seems worthwhile. Note though, that if the compiler fast-fails then this logic will work as before.

Reply:
It's just that with `goto_def_racer_fallback` cfg option enabled, Racer will stop acting as a fallback and both threads will race to return an answer to be used. Since Racer is more of a heuristic compared to analysis data, we might be sacrificing correctness for speed in some cases.

@nrc should chime in about that

Reply:
I would like to preserve the existing behaviour (or at least the intention of the existing behaviour)

* Start both threads at the same time
* If the compiler thread completes, return a result immediately
* if we hit the timeout, use the Racer result if it exists, otherwise return nothing.